CRM/Contact/BAO/Query.php

```diff
@@ -5164,15 +5164,20 @@ public function dateQueryBuilder(
         $date = "'$date'";
       }
 
+      $this->_tables[$tableName] = $this->_whereTables[$tableName] = 1;
+
+      // handle IS NULL / IS NOT NULL / IS EMPTY / IS NOT EMPTY
+      if ($this->nameNullOrEmptyOp($fieldName, $op, $grouping)) {
```

Comment:
Yashodha, line 5167 (in the new numbering) can't be moved earlier than this because not all paths set $this->_tables[$tableName] or $this->_whereTables[$tableName]. Lines 5169-5172 can't be moved to 5066 since there is a path on line 5071 where $this->_where[$grouping] and $this->_qill[$grouping] are not set.

Perhaps I'm being finicky, but I don't know enough about the use of this function to know if your proposal is safe.

However I think the code could be moved to 5139